src/jit/optcse.cpp

```diff
@@ -832,21 +847,94 @@ void Compiler::optValnumCSE_InitDataFlow()
+            //
+            BitVecOps::AddElemD(cseTraits, block->bbCseGen, cseBit);
+            if ((block->bbFlags & BBF_HAS_CALL) == 0)
+            {
+                BitVecOps::AddElemD(cseTraits, block->bbCseGen, cseBit2);
+            }
             lst = lst->tslNext;
         }
     }
 
+    for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+    {
+        // If the block contains a call we may have to update the bbCseGen
+        // as we may generate some CSEs after the last call in the block.
+        //
+        if ((block->bbFlags & BBF_HAS_CALL) != 0)
+        {
+            // We only need to examine blocks that generate CSEs
+            //
+            if (!(BitVecOps::IsEmpty(cseTraits, block->bbCseGen)))
+            {
+                // afterCallGen keeps track of the CSE that are generated after the last call in this block
+                //
+                EXPSET_TP afterCallGen = BitVecOps::MakeEmpty(cseTraits);
+
+                // We walk the statements in this basic block starting at the end and walking backwards,
+                // until we reach the first call
+                //
+                Statement* stmt      = block->lastStmt();
+                bool       foundCall = false;
+                while (!foundCall)
+                {
+                    // Also walk the tree in the backwards direction (bottom up)
+                    // looking for CSE's and updating afterCallGen
+                    // When we reach a call node, we can exit the for loop
+                    //
+                    for (GenTree* tree = stmt->GetRootNode(); tree != nullptr; tree = tree->gtPrev)
+                    {
+                        if (IS_CSE_INDEX(tree->gtCSEnum))
+                        {
+                            unsigned CSEnum  = GET_CSE_INDEX(tree->gtCSEnum);
+                            unsigned cseBit  = genCSEnum2bit(CSEnum) * 2;
+                            unsigned cseBit2 = cseBit + 1;
+
+                            BitVecOps::AddElemD(cseTraits, afterCallGen, cseBit2);
+                        }
+                        if (tree->OperGet() == GT_CALL)
+                        {
+                            // Any cse's that we haven't placed in the afterCallGen set
+                            // aren't currently alive (using cseBit2)
+                            //
+                            foundCall = true;
+                            break;
+                        }
+                    }
+                    // The JIT can sometimes remove the only call in the block
+                    if (stmt == block->firstStmt())
+                    {
+                        break;
+                    }
+                    stmt = stmt->GetPrevStmt();
+                }
+
+                // Set the block->bbCseGen by or-ing in the afterCallGen set
+                //
+                BitVecOps::UnionD(cseTraits, block->bbCseGen, afterCallGen);
```

Comment:
`afterCallGen` doesn't seem to be necessary - it collects CSE defs and then these defs get added to `bbCseGen`. So add the CSE defs directly to `bbCseGen`.

Reply:
OK I can add them directly to bbCseGen instead.